### PR TITLE
use certbot file hooks instead of command line hooks

### LINKED ISSUE #25: [feature request] Custom hook after renewal/creation
It would be nice if we can register/create some script that will be executed after renewal/creation of certificate process.
Example: I'd like to copy certificate to some folder when it's renewed. I know that I can handle this from outside of container with some periodical checking, but if swag contains cron for this, why just don't use that.
So it would be nice if we can have some custom script executed within `--post-hook`. Optionally there could be also custom script for `--pre-hook`.

### LINKED ISSUE #74: le-renew.sh should use --deploy-hook
[linuxserverurl]: https://linuxserver.io
[linuxserver.io][linuxserverurl]

------------------------------

## Expected Behavior

I was looking for a deploy hook of swag found #25 and looked at the hooks of certbot.

After calling certbot -n renew the certificate operations should only be executed when the certificate where actually renewed.
The operations should be done by the --deploy-hook of certbot renew

see [Note ](https://certbot.eff.org/docs/using.html)from certbot:
```
certbot renew exit status will only be 1 if a renewal attempt failed. This means certbot renew 
exit status will be 0 if no certificate needs to be updated. If you write a custom script and expect 
to run a command only after a certificate was actually renewed you will need to use the 
--deploy-hook since the exit status will be 0 both on successful renewal and when renewal is 
not necessary.
```

## Current Behavior

Currently certbot is called like this. This means that all operations are executed within the post-hook.
```
  certbot -n renew \
    --post-hook "if ps aux | grep [n]ginx: > /dev/null; then s6-svc -h /var/run/s6/services/nginx; fi; \
    cd /config/keys/letsencrypt && \
    openssl pkcs12 -export -out privkey.pfx -inkey privkey.pem -in cert.pem -certfile chain.pem -passout pass: && \
    sleep 1 && \
    cat privkey.pem fullchain.pem > priv-fullchain-bundle.pem && \
    chown -R abc:abc /config/etc/letsencrypt"
```

## Steps to Reproduce
See above

## Environment
**OS:** Ubuntu Linux
**CPU architecture:** x86_64
**How docker service was installed:**
official docker repo

## Command used to create docker container (run/create/compose/screenshot)

n/a

## Docker logs

n/a

### PR DESCRIPTION
**Pull request by Feilner**
[linuxserverurl]: https://linuxserver.io
[linuxserver.io][linuxserverurl]

------------------------------

 - [x ] I have read the [contributing](https://github.com/linuxserver/docker-swag/blob/master/.github/CONTRIBUTING.md) guideline and understand that I have made the correct modifications

------------------------------

## Description:
Switch to use certbot file hooks instead of command line hooks.
Fix #25: Custom hooks can now be added as separate persistant files. e.g `/etc/letsencrypt/renewal-hooks/deploy/20-my`
Fix #74: Deploy actions are only done when necessary.

## Benefits of this PR and context:
This PR introduces a more clean usage of hooks by separated files for pre, deploy and post hooks. This make the code more readable (syntax highlighting and indention) and thus better maintainable. It also allows to test the hooks without actually renew the certificate.
The le_renew.sh is now cleaned up only calls `certbot -n renew` which allows a user also easier testing e.g. by calling certbot with out many options.
Last but not least it reduces duplicated code the openssl call and the merge of privkey.pem fullchain.pem could be found actually three times in the code base now it's on a single location. (Actually the merge operation where done in a different way in 50-config. and le-renew.sh `cat {privkey,fullchain}.pem > priv-fullchain-bundle.pem` vs cat `privkey.pem fullchain.pem > priv-fullchain-bundle.pem`

## How Has This Been Tested?

- Build docker container
- Started docker container
- New certificate where created at startup
- Verified that certificate was successful generated and the deploy hook was correctly executed (e.g. priv-fullchain-bundle.pem file was created)
- Verifies that the renew of the certificate works.
- Manual tested the logic in the pre and post hooks. (Modifying /config/.donoteditthisfile.conf and execute the hooks)
exit- Tested to add own hook from external (execute rights are important to set) inside the persistent conf folder `swag_peristant_volume/etc/letsencrypt/renewal-hooks/deploy/20-my`
- Restarted the container to see that hook is still in place
- Logged into the swag docker and manually forced recreation of certificate to see execution of 20-my hook
```
root@swag.fritz.box:/# certbot -n renew --force-renew
Saving debug log to /var/log/letsencrypt/letsencrypt.log

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Processing /etc/letsencrypt/renewal/mydomain.eu.conf
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Plugins selected: Authenticator dns-hetzner, Installer None
Running pre-hook command: /etc/letsencrypt/renewal-hooks/pre/10-nginx
Renewing an existing certificate for *.mydomain.eu and mydomain.eu
Running deploy-hook command: /etc/letsencrypt/renewal-hooks/deploy/10-default
Running deploy-hook command: /etc/letsencrypt/renewal-hooks/deploy/20-my
Output from deploy-hook command 20-my:
Test output from my deploy hook!


- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
new certificate deployed without reload, fullchain is
/etc/letsencrypt/live/mydomain.eu/fullchain.pem
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Congratulations, all renewals succeeded:
  /etc/letsencrypt/live/mydomain.eu/fullchain.pem (success)
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Running post-hook command: /etc/letsencrypt/renewal-hooks/post/10-nginx
```

## Source / References:
[Certbot Hooks](https://certbot.eff.org/docs/using.html#pre-and-post-validation-hooks)
`You can also specify hooks by placing files in subdirectories of Certbot’s configuration directory. Assuming your configuration directory is /etc/letsencrypt, any executable files found in /etc/letsencrypt/renewal-hooks/pre, /etc/letsencrypt/renewal-hooks/deploy, and /etc/letsencrypt/renewal-hooks/post will be run as pre, deploy, and post hooks respectively when any certificate is renewed with the renew subcommand. These hooks are run in alphabetical order and are not run for other subcommands. (The order the hooks are run is determined by the byte value of the characters in their filenames and is not dependent on your locale.)`

For executing the deploy hooks in `50-config` the command `run-parts` is used. This is part of the BusyBox.
[BusyBox](https://busybox.net/downloads/BusyBox.html)

### CODE REVIEW

**Review by tobbenb — CHANGES_REQUESTED** [review]
Looks like you have not read the contribution guidelines. 
You need to edit the readme-vars.yml file to make changes to the readme.

For the rest of the PR, I'm not the right person to approve it or not.

**Comment by Feilner** [comment]
@tobbenb Thanks for the hint I have actually missed this particular part even as clearly noted.
I have reverted changes in README.md and added documentation within readme-vars.yml. I was unsure which date I have to put into the change log as I do not know when it will be merged into the master. (What it hopefully will.)

**Comment by aptalca** [comment]
I appreciate the effort that went into this, but as it is, it's not going to work, because `/etc/letsencrypt` is just a symlink to the `/config/etc/letsencrypt` and that folder is considered ephemeral and gets wiped during cert recreation.

Plus, I'm not really in favor of letting users create custom hooks in an official way, simply because it would add too much support  burden on our team, something we're not equipped to handle with our existing resources.

I try very hard to keep this image as KISS compliant as possible and am very risk averse. Surely there are many things in there that look ugly and complicated; and may not be the `proper` way to do things, but since so many people are using it in production (including myself, where I host various company sites through this image), I adhere to `if it ain't broke, don't fix it`

File: root/etc/cont-init.d/50-config [hunk 1]
```diff
@@ -35,11 +35,19 @@ done
 
 # make our folders and links
 mkdir -p \
-	/config/{log/letsencrypt,log/fail2ban,etc/letsencrypt,fail2ban,crontabs,dns-conf,geoip2db} \
+	/config/{log/letsencrypt,log/fail2ban,etc/letsencrypt,fail2ban,crontabs,dns-conf,geoip2db,renewal-hooks} \
 	/var/run/fail2ban
 rm -rf /etc/letsencrypt
 ln -s /config/etc/letsencrypt /etc/letsencrypt
 
+rm -rf /config/etc/letsencrypt/renewal-hooks
```
Comment (Feilner) [inline root/etc/cont-init.d/50-config:43]: This rm with sub directories covers the update case. As certbot creates the directories renewal-hooks{pre,post,deploy} when it does not exist.

File: root/etc/cont-init.d/50-config [hunk 2]
```diff
@@ -35,11 +35,19 @@ done
...
+ln -s ../../renewal-hooks /config/etc/letsencrypt/renewal-hooks
```
Comment (Feilner) [inline root/etc/cont-init.d/50-config:44]: create relative link to the renewal hooks at `/config/renewal-hooks` due to the relative link it works as well from inside as from outside of the docker.

File: root/etc/cont-init.d/50-config [hunk 3]
```diff
@@ -35,11 +35,19 @@ done
...
+
+# copy default renewal hooks
+chmod -R +x /defaults/renewal-hooks
```
Comment (Feilner) [inline root/etc/cont-init.d/50-config:47]: chmod is only on the default hooks at `/defaults/renewal-hooks` not on the copied `/config/renewal-hooks/` in order to not change behaviour when docker is restarted.

File: root/etc/cont-init.d/50-config [hunk 4]
```diff
@@ -274,6 +282,7 @@ if [ ! "$URL" = "$ORIGURL" ] || [ ! "$SUBDOMAINS" = "$ORIGSUBDOMAINS" ] || [ ! "
   [[ -f /config/etc/letsencrypt/live/"$ORIGDOMAIN"/fullchain.pem ]] && certbot revoke --non-interactive --cert-path /config/etc/letsencrypt/live/"$ORIGDOMAIN"/fullchain.pem --server $REV_ACMESERVER
   rm -rf /config/etc/letsencrypt
   mkdir -p /config/etc/letsencrypt
+  ln -s ../../renewal-hooks /config/etc/letsencrypt/renewal-hooks
```
Comment (Feilner) [inline root/etc/cont-init.d/50-config:285]: restore of symbolic link to renewal-hooks

**Comment by Feilner** [comment]
@aptalca I do dot give up too early :-)
Just pushed a new version which ensures that the hooks are kept in place. I have also have added some review comments to explain some code lines.

The certbot file hooks are always executed by certbot already today. In other words the user could use them already. But with the change the custom hooks would not be deleted when changing the certificate parameters (e.g. URL, STAGING). If you still have doubts we/you can remove the hint to use custom hooks from the documentation.

**Give it a change and think about it.** 

_I have now also tested the case of modifying cert parameters (STAGING) where the `/config/etc/letsencrypt` folder is deleted._

**Pull request closed without merging**